### PR TITLE
Pane Navigation & UIController

### LINKED ISSUE #628: GitHub pane navigation broken
Now that the maintainer workflow branch has been merged to master, it's clear that navigation in the GitHub pane is broken: switching branches switches the pane back to the PR list and subsequent navigation with the back/forward/PR buttons is broken.

### PR DESCRIPTION
**Pull request by StanleyGoldman**
Attempting to resolve #628

In `UIController` When `Jump()` uses `Fire()` to change states, `Fire()` does not correctly determine if the requested state in `requestedTarget` requires an argument.

### CODE REVIEW

File: src/GitHub.VisualStudio/UI/Views/PullRequestListView.xaml.cs [hunk 1]
```diff
@@ -28,7 +28,7 @@ public PullRequestListView()
 
             OpenPR = new RelayCommand(x =>
             {
-                NotifyOpen(new ViewWithData { Data = x });
+                NotifyOpen(new ViewWithData(UIControllerFlow.PullRequests) { ViewType = UIViewType.PRDetail, Data = x});
```
Comment (StanleyGoldman) [inline src/GitHub.VisualStudio/UI/Views/PullRequestListView.xaml.cs]: Based on the unit tests I believe this is required.

File: src/GitHub.App/Controllers/UIController.cs [hunk 2]
```diff
@@ -264,7 +264,8 @@ public void Jump(ViewWithData where)
 
             requestedTarget = where;
             if (activeFlow == where.ActiveFlow)
-                Fire(Trigger.Next, where);
+                //TODO: Fix this
+                Fire(Trigger.Next, where, true);
```
Comment (StanleyGoldman) [inline src/GitHub.App/Controllers/UIController.cs]: Temporary fix to identify a `Fire()` call from `Jump()`

File: src/GitHub.App/Controllers/UIController.cs [hunk 3]
```diff
@@ -799,13 +800,28 @@ bool CreateViewAndViewModel(UIViewType viewType, ViewWithData data = null)
...
         {
             Debug.WriteLine("Firing {0} from {1} ({2})", next, uiStateMachine.State, GetHashCode());
-            if (arg != null && triggers.ContainsKey(next))
-                uiStateMachine.Fire(triggers[next], arg);
+            //TODO: Fix this
+            if (fromJump)
+            {
+                if (requestedTarget.ViewType == UIViewType.PRDetail)
+                {
+                    uiStateMachine.Fire(triggers[triggers.Keys.First()], arg);
```
Comment (StanleyGoldman) [inline src/GitHub.App/Controllers/UIController.cs]: Temporary hard coded fix.

File: src/GitHub.App/Controllers/UIController.cs [hunk 4]
```diff
@@ -802,10 +802,33 @@ bool CreateViewAndViewModel(UIViewType viewType, ViewWithData data = null)
...
+            if (arg != null)
+            {
+                if (next == Trigger.Next)
+                {
+                    switch (requestedTarget?.ViewType)
+                    {
+                        case UIViewType.PRDetail:
+                            viewWithDataTrigger = triggers[Trigger.PRDetail];
+                            break;
+                    }
```
Comment (StanleyGoldman) [inline src/GitHub.App/Controllers/UIController.cs]: Checking if the `requestedTarget` needs a parameter.

File: src/GitHub.App/Controllers/UIController.cs [hunk 5]
```diff
@@ -258,9 +258,15 @@ public void Start([AllowNull] IConnection conn)
...
             if (where.ActiveFlow != mainFlow)
-                return;
+            {
+                var message = string.Format(CultureInfo.CurrentCulture,
+                    "Jump called for flow {0} but this is {1}", 
+                    where.ActiveFlow, 
+                    mainFlow);
+
+                throw new ArgumentException(message);
+            }
```
Comment (StanleyGoldman) [inline src/GitHub.App/Controllers/UIController.cs]: I think it's better to throw an error in this situation than `Assert()`.

File: src/UnitTests/GitHub.App/Controllers/UIControllerTests.cs [hunk 6]
```diff
@@ -308,37 +446,71 @@ public void AuthFlowWith2FA()
...
+                    Assert.Equal(testViewCount + 1, count);
+                    count++;
                 });
 
                 uiController.Start(null);
-                Assert.Equal(3, count);
+                Assert.Equal(testViewCount + 2, count);
                 Assert.True(uiController.IsStopped);
+                Assert.True(success.HasValue);
+                Assert.True(success);
```
Comment (StanleyGoldman) [inline src/UnitTests/GitHub.App/Controllers/UIControllerTests.cs]: `Assert.True(success);` bothers me here. If the last action performed was Cancel on the clone flow, shouldn't this value be False?

**Comment by StanleyGoldman** [comment]
All the changes up to this point resolve the "Back and Forth" issue reported in #628, but I'm noticing a bit different behavior on the "Refresh".

**Comment by StanleyGoldman** [comment]
I'm seeing an invalid state transition from PRDetail to PRDetail

**Comment by StanleyGoldman** [comment]
I think the problem is due to the Next Trigger attempting to redirect to PRDetail which is a TriggerWithParam.

My original fix attempted to intercept the Next Trigger in that scenario and issue a PRDetail TriggerWithParam, that works in a few scenarios, but it would require permitting a PRDetail TriggerWithParam everywhere a Next Trigger is specified.

I cleaned up, uniformed  and commented the tests in `UIControllerTests` them while I read them. I added a `Jump` unit test to `PullRequestsFlow` which I believe reproduces the issue at hand.

In `UIController` I changed the `Debug.WriteLine()` calls to `Log.Info()`; and I added additional logs for state machine transitions and unhandled transitions. Reading the log of `UIController` really helped me understand it.

**Comment by StanleyGoldman** [comment]
After not thinking about it for a day, I think one solution could be to make every trigger a TriggerWithParam. Even if a param is not used/required; that would allow the Next operation to work regardless if the "Next" state needs a param or not.

File: src/UnitTests/GitHub.App/Controllers/UIControllerTests.cs (hunk 6 above)
Comment (shana) [inline src/UnitTests/GitHub.App/Controllers/UIControllerTests.cs]: iirc, `success` is `false` if the flow failed at the end without user intervention. The user cancelling a flow would not be a failure of the flow itself, merely an early exit. And yes, it doesn't feel very logical. I need to revisit some of the thoughts on this one...

**Comment by shana** [comment]
To be honest, I think UIController is taking on too much responsibility and shouldn't be handling all this work, which is manifesting itself in all these problems. It's meant to be a state machine handling logic for ui flows, which really doesn't work well when forced to jump to arbitrary points in the middle of a particular flow.

**Comment by grokys** [comment]
Thanks for the attempt @StanleyGoldman, but we decided to rewrite the navigation controller in #752 so closing this.

**Pull request closed without merging**